polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -157,51 +154,72 @@ class PolyKernel private[kernel] (
-                    case result =>
-                      IO.pure(result)
-                  }
-                }
-
-                for {
-                  _ <- symbolTable.drain()
-                  start <- clock.monotonic(MILLISECONDS)
-                  res <- results
-                } yield {
-                  res.onComplete {
-                    Stream.eval {
-                      for {
-                        end <- clock.monotonic(MILLISECONDS)
-                        timestamp <- clock.realTime(MILLISECONDS)
-                      } yield ExecutionInfo((end - start).toInt, timestamp)
+                CellContext(cell.id).flatMap {
+                  cellContext =>
+                    IO(notebookContext.tryUpdate(cellContext)).flatMap {
+                      prevContext =>
+                        val results = interp.runCode(
+                          cellContext,
+                          cell.content.toString
+                        ).map {
+                          results => results.through(cellContext.results.tapResults).evalTap {
+                            case ResultValue(name, _, _, _, value, _) => IO(polynote.runtime.Runtime.putValue(name, value))
+                            case _ => IO.unit
+                          }
+                        }.handleErrorWith {
+                          err => IO(prevContext.map(notebookContext.tryUpdate)) *> IO.raiseError(err) // restore previous context on error
+                        }
+
+                        for {
+                          start  <- clock.monotonic(MILLISECONDS)
+                          res    <- results
+                        } yield {
+                          res.onComplete {
+                            Stream.eval {
+                              for {
+                                end <- clock.monotonic(MILLISECONDS)
+                                timestamp <- clock.realTime(MILLISECONDS)
+                              } yield ExecutionInfo((end - start).toInt, timestamp)
+                            }
+                          }.onFinalize {
+                            // if the interpreter didn't make a module for the cell, use the scala interpreter to make one
+                            cellContext.module.tryGet.flatMap {
+                              case Some(_) => IO.unit
+                              case None =>
+                                getInterpreter("scala") match {
+                                  case Some(interp: ScalaInterpreter) =>
+                                    import interp.kernelContext.global.Quasiquote
+                                    val exports = cellContext.resultValues.foldLeft(SortedMap.empty[String, interp.kernelContext.global.Type]) {
+                                      (accum, rv) => accum + (rv.name -> rv.scalaType.asInstanceOf[interp.kernelContext.global.Type])
+                                    }.toList.map {
+                                      case (name, typ) =>
```

Comment:
Basically if the interpreter didn't generate a classfile we're generating this one for them:

```scala
object Cell2$Eval$6 {  // or whatever cells get named
  val foo: String = polynote.runtime.Runtime.getValue("foo").asInstanceOf[String]
  val bar: Int = polynote.runtime.Runtime.getValue("bar").asInstanceOf[Int]
}
```
etc. This might run into the same issue as we had when Scala cells were simple objects; I'm not sure. But it should be far less prevalent and is much simpler for now. (and is probably no more broken in Spark cases than the previous mechanism)

Reply:
And, this object is then initialized so that the values are immediately grabbed. They'll stay frozen in that object even if they get shadowed by future definitions! Reproducibility!

Reply:
Nice!